### PR TITLE
Contraction path node

### PR DESCRIPTION
**Pull request by mganahl**
Add `contract_path` and `path_solver` for use with `Node`

### CODE REVIEW

**Comment by mganahl** [comment]
Hey @Thenerdstation pls have a look at this. This is something that we needed for a long time. Tests will be added after your had a chance to look at this. This should be pulled in ASAP because I'd make it part of a presentation

File: tensornetwork/contractors/opt_einsum_paths/path_contractors.py [hunk 1]
```diff
@@ -290,3 +290,82 @@ def custom(nodes: Iterable[AbstractNode],
...
+  elif algorithm == "branch":
+    alg = functools.partial(
+      opt_einsum.paths.branch, memory_limit=memory_limit, nbranch=nbranch)
+  elif algorithm == "greedy":
+    alg = functools.partial(opt_einsum.paths.greedy, memory_limit=memory_limit)
+  elif algorithm == "auto":
+    n = len(list(nodes))  #pytype thing
+    _nodes = nodes
+    if n <= 1:
+      raise ValueError(f"network is trivial with {n} nodes. ")
```
Comment (Thenerdstation) [inline tensornetwork/contractors/opt_einsum_paths/path_contractors.py:325]: Prefer to just return a trivial empty path.

**Review by Thenerdstation — APPROVED** [review]
Looks good. Only concern is passing in the algorithm as a string feels dirty, but I don't expect us to add a ton more so this should be fine for now.

**Comment by Thenerdstation** [comment]
Also yes please add tests

**Comment by mganahl** [comment]
> Looks good. Only concern is passing in the algorithm as a string feels dirty, but I don't expect us to add a ton more so this should be fine for now.
It's not beautiful, but the advantage is that the user doesn't have to pull in a module for that.

**Comment by Thenerdstation** [comment]
> It's not beautiful, but the advantage is that the user doesn't have to pull in a module for that.

Agreed. Simplicity is a virtue.

**Comment by Thenerdstation** [comment]
Test + passing build and I'll merge this in.

If you also want, I can also push version 0.4.3 once this is in

**Comment by mganahl** [comment]
> If you also want, I can also push version 0.4.3 once this is in

That would be great!

**Pull request merged**